### ISSUE TITLE
libcurl.so.4: no version information

### ISSUE THREAD

**Issue by amper5and** [Blitznote/docker-ubuntu-debootstrap#10]
I'm using `blitznote/debootstrap-amd64:17.04`. After installing a package that uses libcurl, I get the following error message every time my program uses libcurl:
```
 /usr/lib/x86_64-linux-gnu/libcurl.so.4: no version information available (required by .... )
```
It works, but it first prints out that message *every single time.*

Do you think this is because you are compiling your libraries without version information?

**Comment by mark-kubacki** [comment]
That is actually not the case, and version information is included in the library.

Please check with the author(s) of whatever it is that links to *libcurl*. Perhaps an obsolete API call or understanding of the version format is the cause?

Please let me know If there's something I can do. As I don't strip any symbols from *curl* I am closing this for now.

**Comment by amper5and** [comment]
Thanks for your comments and for confirming that version information is included. I agree likely an issue with the binary calling libcurl. 

And thank you for these nice minimal Ubuntu images!

**Comment by mark-kubacki** [comment]
Binaries from packages like *apt* and *git* (as well as my *PHP* images) use *libcurl*, and neither during compilation nor use did I notice aforementioned warning.

Clicking through logs on my CI server the message shows up with *cmake* and the *Steam* client. I will look into these pairings, though with a low priority because their result is the same with Ubuntu's flavour installed. (I just checked it.)

This is how you would switch to Ubuntu's packages only (remove `-s`), mind the trailing *minus*:
```bash
. /etc/os-release
apt-get -s --allow-downgrades install \
  curl/${UBUNTU_CODENAME} \
  apt-transport-https/${UBUNTU_CODENAME} \
  libcurl3-boringssl-
```

**Comment by amper5and** [comment]
For reference, the binaries that give me this warning are from [megatools](https://packages.ubuntu.com/zesty/megatools), installed from `universe` with *apt*.

**Comment by mark-kubacki** [comment]
For my notes: `debug-missing-symbols-warning@packages/curl`

**Comment by mark-kubacki** [comment]
`readelf -V /usr/lib/x86_64-linux-gnu/libcurl-boringssl.so.4`  
… shows there's no section `.gnu.version_d` in my builds.

Configuring with `--enable-versioned-symbols` results in a later `LDFLAGS_EXTRA += -Wl,--version-script=libcurl.vers` being used, which gives us this missing bit.

A new version of *curl* is announced for 2017-11-29 (in two days), and I will publish new packages and my container images the same or the day after that. They will contain the fix.

Thanks for bringing this issue to my attention!

**Comment by mark-kubacki** [comment]
Turns out this breaks a lot of Debian's and Ubuntu's packages: They rely on `CURL_OPENSSL_3`—with the `_3` being the relevant part here—and curl already is on `_4`. Debian maintainers chose to stick with the `_3` in export maps.  
https://anonscm.debian.org/cgit/collab-maint/curl.git/tree/debian/patches/03_keep_symbols_compat.patch?h=debian/7.56.1-1

I've added some symbol aliases to work around this.

----
For my notes: https://www.akkadia.org/drepper/dsohowto.pdf

**Comment by mark-kubacki** [comment]
Much to my surprise this in turn breaks another array of binaries, when building from source. I learned that there's an arcane symbol lookup order. Fortunately this article from @ianlancetaylor outlines it (thanks!):  
https://www.airs.com/blog/archives/300

I've reordered (and duplicated) symbols accordingly, and for now all tests pass. Should something come I will duplicate the files with an old and new version section, and merge the common parts to save space again (→ reflinks).
